### PR TITLE
Stop auto-followers when resetting metadata

### LINKED ISSUE #81242: Auto-follow pattern and remote might not be stopped when restoring from old snapshot with global state
**Elasticsearch version** (`bin/elasticsearch --version`): 7.16

**Description of the problem including expected versus actual behavior**:

According to the code:

https://github.com/elastic/elasticsearch/blob/0f108251341303dbf4afabaadc8f1d673a368b0c/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java#L244-L248

AutoFollowCoordinator will not perform any actions (such as stopping already running patterns and remotes) if the new auto-follow metadata is null. This could theoretically happen when restoring from a snapshot with global state that was taken **before** any auto-follower patterns were configured or even before the feature were available.

**Steps to reproduce**:

* start a new cluster
* configure a snapshot repository
* create a snapshot (with `include_global_state=true`)
* configure remote and auto-follow pattern
* restore from the snapshot (with `include_global_state=true`)

Expected result:
* existing auto-follow patterns are stopped

Actual result:
* auto-follow patterns are not stopped

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change ensures auto-followers are stopped if auto-follow metadata is reverted back to null after some non-empty state.

Fixes #81242 (though will not happen until #81247 is fixed)

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java [hunk 1]
```diff
@@ -241,10 +241,8 @@ synchronized void updateStats(List<AutoFollowResult> results) {
     }
 
     void updateAutoFollowers(ClusterState followerClusterState) {
-        final AutoFollowMetadata autoFollowMetadata = followerClusterState.getMetadata().custom(AutoFollowMetadata.TYPE);
-        if (autoFollowMetadata == null) {
-            return;
-        }
+        final AutoFollowMetadata autoFollowMetadata = followerClusterState.getMetadata()
+            .custom(AutoFollowMetadata.TYPE, AutoFollowMetadata.EMPTY);
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java:246]: Can we bail out early here if `autoFollowMetadata.getPatterns().isEmpty() && this.autoFollowers.isEmpty()`? Since this runs on every cluster change it would be nice to have a bail out.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java [hunk 2]
```diff
@@ -1147,25 +1147,43 @@ public void testUpdateAutoFollowers() {
...
+        );
+        var initialAutoFollowers = autoFollowCoordinator.getAutoFollowers();
+
+        // when resetting the state
+        autoFollowCoordinator.updateAutoFollowers(createClusterStateWith(null));
+        var newAutoFollowers = autoFollowCoordinator.getAutoFollowers();
+
+        // then auto-followers are removed
+        assertThat(newAutoFollowers.entrySet(), empty());
+        // and remotes are stopped
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java:1169]:
```suggestion
        // and auto-followers are stopped
```

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java [hunk 3]
```diff
@@ -1147,25 +1147,43 @@ public void testUpdateAutoFollowers() {
...
+        assertThat(autoFollowCoordinator.getAutoFollowers().keySet(), empty());
         assertThat(autoFollowCoordinator.getStats().getAutoFollowedClusters().size(), equalTo(0));
     }
 
     public void testUpdateAutoFollowersNoAutoFollowMetadata() {
-        AutoFollowCoordinator autoFollowCoordinator = createAutoFollowCoordinator();
-        ClusterState clusterState = ClusterState.builder(new ClusterName("remote")).build();
-        autoFollowCoordinator.updateAutoFollowers(clusterState);
+        var autoFollowCoordinator = createAutoFollowCoordinator();
+        autoFollowCoordinator.updateAutoFollowers(createClusterStateWith(null));
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java:1184]: Perhaps randomly choose `null` or `Map.of()`?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java (hunk 3 above)
Comment (idegtiarenko) [reply x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java:1184]: There is already another test case that checks transition from non empty to empty map.

**Comment by idegtiarenko** [comment]
@elasticmachine run elasticsearch-ci/bwc
